src/jaxsim/rbda/contacts/visco_elastic.py

```diff
@@ -0,0 +1,1050 @@
+
+        # Compute the Jacobians of the material deformation rate w.r.t. the state.
+        dṁ_dp = jnp.vstack(dṁ_dp_fun(p0, v0, m0))
+        dṁ_dv = jnp.vstack(dṁ_dv_fun(p0, v0, m0))
+        dṁ_dm = jnp.vstack(dṁ_dm_fun(p0, v0, m0))
+
+        # Pack the A matrix.
+        A_sc = jnp.vstack(
+            [
+                jnp.hstack([df_dp, df_dv, df_dm]),
+                jnp.hstack([dṁ_dp, dṁ_dv, dṁ_dm]),
+            ]
+        )
+
+        # ============
+        # Compute b_sc
+        # ============
+
+        # Compute the output of the non-linear model at the initial state.
+        x0 = jnp.hstack([p0, v0, m0])
+        f0, ṁ0 = compute_contact_force_non_linear_model(
+            position=p0, velocity=v0, tangential_deformation=m0
+        )
+
+        # Pack the b vector.
+        b_sc = jnp.hstack([f0, ṁ0]) - A_sc @ x0
+
+        return A_sc, b_sc
+
+    @staticmethod
+    @functools.partial(jax.jit, static_argnames=("terrain",))
+    def compute_contact_force_non_linear_model(
+        position: jtp.VectorLike,
+        velocity: jtp.VectorLike,
+        tangential_deformation: jtp.VectorLike,
+        parameters: ViscoElasticContactsParams,
+        terrain: Terrain,
+    ) -> tuple[jtp.Vector, jtp.Vector]:
+        """
+        Compute the contact forces using the non-linear Hunt/Crossley model.
+
+        Args:
+            position: The position of the contact point.
+            velocity: The velocity of the contact point.
+            tangential_deformation: The tangential deformation of the contact point.
+            parameters: The parameters of the contact model.
+            terrain: The considered terrain.
+
+        Returns:
+            A tuple containing:
+            - The linear contact force in the mixed contact frame.
+            - The rate of material deformation.
+        """
+
+        # Compute the linear contact force in mixed representation using
+        # the non-linear Hunt/Crossley model.
+        # The following function also returns the rate of material deformation.
+        CW_fl, ṁ = SoftContacts.hunt_crossley_contact_model(
+            position=position,
+            velocity=velocity,
+            tangential_deformation=tangential_deformation,
+            terrain=terrain,
+            K=parameters.K,
+            D=parameters.D,
+            mu=parameters.static_friction,
+            p=parameters.p,
+            q=parameters.q,
+        )
+
+        return CW_fl, ṁ
+
+    @staticmethod
+    @jax.jit
+    def integrate_data_with_average_contact_forces(
+        model: js.model.JaxSimModel,
+        data: js.data.JaxSimModelData,
+        *,
+        dt: jtp.FloatLike,
+        link_forces: jtp.MatrixLike | None = None,
+        joint_force_references: jtp.VectorLike | None = None,
+        average_link_contact_forces_inertial: jtp.MatrixLike | None = None,
+        average_of_average_link_contact_forces_mixed: jtp.MatrixLike | None = None,
```

Comment:
Since this is called only internally, these arguments should be `jnp.array` already, right?
```suggestion
        link_forces: jtp.Matrix | None = None,
        joint_force_references: jtp.Vector | None = None,
        average_link_contact_forces_inertial: jtp.Matrix | None = None,
        average_of_average_link_contact_forces_mixed: jtp.Matrix | None = None,
```

Reply:
This method, as it has been originally though, is a particular implementation of a forward Euler scheme where the $\bar{\mathbf{f}}$ and $\bar{\bar{\mathbf{f}}}$ are available. In theory, these averages can be computed with different methodologies, including the exponential integration. For this reason, this method is marked as public and can operate on `*Like` types.

---

I have never investigated if wrapping a jax array with another jax array in a jit-compiled method causes any overhead. At first look, the following behave the same if the dtype of `x` is already correct:

```python
@jax.jit
def func1(x):
    return x * x

@jax.jit
def func2(x):
    x = jnp.array(x, dtype=float)
    return x * x

lowered1 = func1.lower(jnp.array(0.0))
lowered2 = func2.lower(jnp.array(0.0))

compiled1 = lowered1.compile()
compiled2 = lowered2.compile()

print(compiled1.cost_analysis())
# [{'utilization0{}': 1.0, 'bytes accessedout{}': 8.0, 'bytes accessed': 24.0, 'flops': 1.0, 'bytes accessed0{}': 8.0, 'utilization1{}': 1.0, 'bytes accessed1{}': 8.0}]

print(compiled2.cost_analysis())
# [{'utilization0{}': 1.0, 'bytes accessedout{}': 8.0, 'bytes accessed': 24.0, 'flops': 1.0, 'bytes accessed0{}': 8.0, 'utilization1{}': 1.0, 'bytes accessed1{}': 8.0}]
```

In light of this preliminary analysis, it seems ok to operate when possible on `*Like` arguments since they 1) casting does not introduce overhead if they are already of the right type, and 2) the work also on non-JAX-array types. Of course, if a function is private and we already know that we pass a jax array, we can omit the cast for compactness.

Reply:
Thank you for investigating! I agree with you, re-wrapping a `jnp.array` shouldn't be a problem, judging from [`jax._src.numpy.lax_numpy.array`](https://github.com/jax-ml/jax/blob/023f2a78be4bcdd79fc7e7b8c60c1161155e8928/jax/_src/numpy/lax_numpy.py#L4950-L5119)